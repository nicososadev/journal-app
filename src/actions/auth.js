import { firebase, googleAuthProvider } from '../firebase/firebase-config';
import { types } from '../types/types'

export const loginManager = (email, password) => {

    return (dispatch) => {

        setTimeout(() => {
            
            dispatch( login(123, 'Juan') )
        }, 3500);
    }
}

export const googleLoginManager = () => {
    
    return (dispatch) => {

        firebase.auth().signInWithPopup(googleAuthProvider)
            .then( ({ user }) => {
                dispatch(
                    login(user.uid, user.displayName)
                )
            })
    }
}

export const login = (uid, displayName) => {

    return {
        type: types.login,
        payload: {
            uid,
            displayName
        }
    }
}