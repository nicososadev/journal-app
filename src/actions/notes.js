import Swal from "sweetalert2"
import { db } from "../firebase/firebase-config"
import { fileUpload } from "../helpers/fileUpload"
import { loadNotes } from "../helpers/loadNotes"
import { types } from "../types/types"

export const createNote = () => {

    return async(dispatch, getState) => {

        const state = getState()

        const { uid } = state.auth

        const newNote = {
            title: '',
            body: '',
            date: new Date().getTime()
        }

        const document = await db.collection(`${uid}/journal/notes`).add(newNote)

        dispatch( activeNote(document.id, newNote) )

        dispatch( createNewNote(document.id, newNote) )

    }
}

export const activeNote = (id, note) => {

    return {
        type: types.notesActive,
        payload: {
            id,
            ...note
        }
    }
}

export const createNewNote = (id, note) => {

    return {
        type: types.notesCreateNote,
        payload: {
            id,
            ...note
        }
    }
}

export const LoadNotesManager = (uid) => {

    return async (dispatch) => {
        
        const notes = await loadNotes( uid )

        dispatch( setNotes(notes) )
    }
}

export const setNotes = (notes) => {

    return {
        type: types.notesLoad,
        payload: notes
    }
}

export const saveNoteManager = ( note ) => {

    return async (dispatch, getState) => {
        
        const state = getState()

        const { uid } = state.auth

        if( !note.url ){
            delete note.url
        }

        const noteToFireStore = { ...note }

        delete noteToFireStore.id

        await db.doc(`${uid}/journal/notes/${note.id}`).update( noteToFireStore )

        dispatch( noteRefresh( note.id, noteToFireStore ) )

        Swal.fire('Saved', note.title, 'success')
    }
}

export const noteRefresh = ( id, note ) => {

    return {
        type: types.notesUpdateNote,
        payload: {
            id, 
            note: {
                id,
                ...note
            }
        }
    }
}

export const uploadManager = ( file ) => {

    return async (dispatch, getState) => {

        const { active:activeNote } = getState().notes

        Swal.fire({
            title: 'Uploading...',
            text: 'Please wait...',
            allowOutsideClick: false,
            willOpen: () => {
                Swal.showLoading()
            }
        })
        

        const fileUrl = await fileUpload( file )

        activeNote.url = fileUrl

        dispatch( saveNoteManager( activeNote ) )

        Swal.close()
    }
}

export const deleteManager = ( id ) => {

    return async (dispatch, getState) => {

        const state = getState()

        const { uid } = state.auth

        await db.doc(`${uid}/journal/notes/${id}`).delete()

        dispatch( deleteNote( id ) )
    }
}

export const deleteNote = ( id ) => {
    return {
        type: types.notesDeleteNote,
        payload: id
    }
}

export const notesLogout = () => {
    return{
        type: types.notesLogoutCleaning
    }
}