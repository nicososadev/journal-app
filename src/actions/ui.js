import { types } from "../types/types"

export const setError = ( error ) => {
   
    return {
        type: types.uiSetError,
        payload: error
    }
}
