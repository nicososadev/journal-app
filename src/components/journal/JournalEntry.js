import React from 'react'

export const JournalEntry = () => {
    
    return (
        <div className="journal__entry pointer">

            <div className="journal__entry-picture">
            </div>
            
            <div className="journal__entry-body">
                <p className="jounral__entry-title">Titulo</p>
                <p className="jounral__entry-content">Contenido del Entry</p>
            </div>
            
            <div className="journal__entry-date">
                <span>Monday</span>
                <h4>28</h4>
            </div>
        </div>
    )
}
