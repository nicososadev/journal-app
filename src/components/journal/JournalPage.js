import React from 'react'

export const JournalPage = () => {
    return (
        <div>
            <h1>Journal Page</h1>
        </div>
    )
}
