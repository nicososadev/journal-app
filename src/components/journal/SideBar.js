import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { logoutManager } from '../../actions/auth'
import { createNote } from '../../actions/notes'
import { JournalEntries } from './JournalEntries'

export const SideBar = () => {

    const dispatch = useDispatch()

    const { name } = useSelector(state => state.auth)

    const handleLogout = () => {

        dispatch( logoutManager() )
    }

    const handleNewEntry = () => {
        dispatch( createNote() )
    }

    return (

        <aside className="journal__sidebar">
            <div className="journal__sidebar-navbar">
                <h3 className="mt-5">
                    <i className="far fa-moon" />
                    <span> {name}</span>
                </h3>

                <button onClick={handleLogout} className="btn">Logout</button>
            </div>

            <div onClick={handleNewEntry} className="journal__new-entry">
                <i className="far fa-calendar-plus fa-5x" />
                <p className="mt-5">New Entry</p>
            </div>

            <JournalEntries />
        </aside>
    )
}
