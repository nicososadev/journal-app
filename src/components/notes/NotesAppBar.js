import React from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { saveNoteManager } from '../../actions/notes'

export const NotesAppBar = () => {

    const dispatch = useDispatch()

    const { active } = useSelector(state => state.notes)

    const handleSaveNote = () => {

        dispatch( saveNoteManager( active ) )
    }

    return (

        <div className="notes__appbar">
            <span>28 de agosto 2020</span>
            <div>
                <button className="btn">Picture</button>
                <button onClick={handleSaveNote} className="btn">Save</button>
            </div>
        </div>
    )
}
