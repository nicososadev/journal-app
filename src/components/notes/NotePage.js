import React, { useEffect, useRef } from 'react'
import { useDispatch, useSelector } from 'react-redux'
import { activeNote, deleteManager } from '../../actions/notes'
import { useForm } from '../../hooks/useForm'
import { NotesAppBar } from './NotesAppBar'

export const NotePage = () => {

    const dispatch = useDispatch()

    const { active:note } = useSelector(state => state.notes)

    const [ formValues, handleInputChange, reset ] = useForm( note )

    const { title, body } = formValues

    const activeId = useRef( note.id )

    useEffect(() => {

        if( note.id !== activeId.current ){
            reset( note )
            activeId.current = note.id
        }

    }, [note, reset])

    useEffect(() => {
        
        dispatch( activeNote( formValues.id, { ...formValues } ) )

    }, [formValues, dispatch])

    const handleDelete = () => {

        dispatch( deleteManager( activeId.current ) )
    }

    return (

        <div className="notes__main-content">

            <NotesAppBar />

            <div className="notes__content">
                <input type="text" className="notes__title-input" value={title} name="title" onChange={handleInputChange} placeholder="Some awesome title" />
                <hr className="mb-5" />
                <textarea className="notes__textarea" value={body} name="body" onChange={handleInputChange} placeholder="An awesome text">

                </textarea>
                
                {
                    (note.url) &&
                    <div className="notes__image">
                        <img src={note.url} alt="imagen"/>
                    </div>
                }
            </div>

            <button
                className="btn btn-danger"
                onClick={handleDelete}
            >
                Delete
            </button>
        </div>

    )
}
