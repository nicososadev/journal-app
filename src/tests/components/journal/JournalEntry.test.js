import '@testing-library/jest-dom'
import React from 'react'
import thunk from 'redux-thunk'
import configureStore from 'redux-mock-store'
import { mount } from 'enzyme'
import { Provider } from 'react-redux'
import { JournalEntry } from '../../../components/journal/JournalEntry'
import { activeNote } from '../../../actions/notes'

const middlewares = [thunk]
const mockStore = configureStore(middlewares)

const initialState = {}

let store = mockStore(initialState)

store.dispatch = jest.fn()

describe('Pruebas en el componente JournalEntry', () => {
    
    const note = {
        id: 1234,
        body: 'Nota de Prueba',
        title: 'Prueba',
        date:0,
        url: 'https://url.com/img.jpg'
    }

    const wrapper = mount(
        <Provider store={store}>
            <JournalEntry {...note} />
        </Provider>
    )

    test('Debe mostrarse correctamente', () => {
        
        expect( wrapper ).toMatchSnapshot()
    })

    test('Debe llamar a la accion activeNote', () => {
        
        wrapper.find('.journal__entry').simulate('click')

        expect( store.dispatch ).toHaveBeenCalledWith(
            activeNote( note.id, {...note})
        )
    })
})
