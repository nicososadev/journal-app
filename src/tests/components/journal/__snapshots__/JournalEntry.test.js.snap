// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Pruebas en el componente JournalEntry Debe mostrarse correctamente 1`] = `
<div
  className="journal__entry pointer animate__animated animate__fadeIn animate__faster"
  onClick={[Function]}
>
  <div
    className="journal__entry-picture"
    style={
      Object {
        "backgroundImage": "url(https://url.com/img.jpg)",
        "backgroundSize": "cover",
      }
    }
  />
  <div
    className="journal__entry-body"
  >
    <p
      className="jounral__entry-title"
    >
      Prueba
    </p>
    <p
      className="jounral__entry-content"
    >
      Nota de Prueba
    </p>
  </div>
  <div
    className="journal__entry-date"
  >
    <span>
      Wednesday
    </span>
    <h4>
      31st
    </h4>
  </div>
</div>
`;
